Update lint rules, and clarify docs around linting

ttmarek opened the pull request:
@Timer I made a long-winded issue a couple of days ago about linting. I've closed that issue because it was mostly rant. Instead, I'm proposing some changes to the ESLint rules and docs that we can iterate on in this pull request.

1. I updated CRA's rules so they are all configured as warnings.
2. I removed the ESLint rules that are style-specific. I did this by looking at rules that we had that were in common with those in `eslint-config-prettier`.
3. I updated the docs to clarify how linting works in CRA, and to provide some tips on how to do certain things that I struggled with initially.

gaearon commented:
>I updated CRA's rules so they are all configured as warnings.

Where can I read the context on this? Very few rules (two or three) are intentionally errors because they protect against confusing rabbit holes where beginners would spend more time than learning how to fix the error.

Timer commented:
See https://github.com/facebookincubator/create-react-app/issues/3315 for some context.

I believe he's referring to https://github.com/ttmarek/create-react-app/blob/101366c3bc02cc7dd00b1cc4b86d154735b2eee0/packages/eslint-config-react-app/index.js#L13-L14.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -241,15 +224,15 @@ module.exports = {
-    'import/no-webpack-loader-syntax': 'error',
+    'import/first': 'warn',
+    'import/no-amd': 'warn',
+    'import/no-webpack-loader-syntax': 'warn',
```
These should stay errors. They're intentionally breaking the build for features we don't support.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -241,15 +224,15 @@ module.exports = {
-    'import/first': 'error',
-    'import/no-amd': 'error',
-    'import/no-webpack-loader-syntax': 'error',
+    'import/first': 'warn',
```
This should stay an error. People often misunderstand that ES imports are hoisted, and try to do side effects between them. It's very confusing that it doesn't work. This lint rule altogether prevents this from happening and trains you to write imports in the way the bundler actually interprets them.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -241,15 +224,15 @@ module.exports = {
     'react/jsx-no-duplicate-props': ['warn', { ignoreCase: true }],
     'react/jsx-no-target-blank': 'warn',
-    'react/jsx-no-undef': 'error',
+    'react/jsx-no-undef': 'warn',
```
This will always be a runtime error. Guaranteed. So there's no need to wait until the code executes. Please keep it an error.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -174,8 +160,8 @@ module.exports = {
     'no-throw-literal': 'warn',
-    'no-undef': 'error',
-    'no-restricted-globals': ['error'].concat(restrictedGlobals),
+    'no-undef': 'warn',
```
This should intentionally be an error. We want to train people to qualify globals with `window.` when they mean to use a global. It's not a stylistic choice, but to clearly separate the concept of npm dependencies from a concept of the globals.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -174,8 +160,8 @@ module.exports = {
-    'no-undef': 'error',
-    'no-restricted-globals': ['error'].concat(restrictedGlobals),
+    'no-undef': 'warn',
+    'no-restricted-globals': ['warn'].concat(restrictedGlobals),
```
This, too, should stay an error. If you write code like

```js
handleChange() {
  this.setState({text: event.target.value});
}
```

your build should break. Otherwise you'll spend hours chasing this accidental global usage.

It's just not worth it to allow the build.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -215,16 +201,13 @@ module.exports = {
     'valid-typeof': 'warn',
     'no-restricted-properties': [
-      'error',
+      'warn',
```
Should stay an error. It's not about style but about the subset of features we promise to support.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -263,8 +246,8 @@ module.exports = {
-    'react/react-in-jsx-scope': 'error',
-    'react/require-render-return': 'error',
+    'react/react-in-jsx-scope': 'warn',
+    'react/require-render-return': 'warn',
```
If you don't return it will always blow up. Should break the build.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -263,8 +246,8 @@ module.exports = {
     'react/no-is-mounted': 'warn',
-    'react/react-in-jsx-scope': 'error',
-    'react/require-render-return': 'error',
+    'react/react-in-jsx-scope': 'warn',
```
If you don't do this it will always blow up at runtime. Should stay an error.

ttmarek commented:
Thanks for the review @gaearon. Yeah @Timer gave me a good explanation for why some of the eslint rules are errors. I just haven't had a chance to update this PR since then. I will remove those changes.

How do you guys feel about the other changes though? Removing certain style-specific eslint rules and updating the docs around linting?

gaearon commented:
Removing style rules seems fine.
I haven't read the doc changes yet.

gaearon commented on packages/eslint-config-react-app/index.js:
```diff
@@ -10,8 +10,6 @@
 // Inspired by https://github.com/airbnb/javascript but less opinionated.
 
 // We use eslint-loader so even warnings are very visible.
-// This is why we only use "WARNING" level for potential errors,
```
Instead of removing it would be better to clarify the criteria we use to decide which it should be

ttmarek replied:
Yeah, fair enough. Umm...So from what I understand, we have some eslint rules that are configured as errors to catch run time errors without having to build and run the app.

 - 'react/jsx-no-undef'
 - 'react/react-in-jsx-scope'
 - 'react/require-render-return'

Makes sense that these are errors.

We also have some eslint rules that are configured as errors to prevent users from using features that aren't supported by CRA:

 - 'import/first'
 - 'import/no-amd'
 - 'import/no-webpack-loader-syntax'
 - 'no-restricted-properties'

Makes sense that these are errors too.

But, the following three eslint rules that are configured as errors are a bit different. If I understand correctly these are in place to help educate or train CRA users to follow certain best practices. But couldn't the same be said for every other eslint rule we have? I guess what I'm wondering, is how can we justify that these are set as errors while other eslint rules are set as warnings. Is it the severity of not following these practices? Or the difficultly of tracking bugs that could arise from not following these practices?

 - 'import/first'
 - 'no-restricted-globals'
 - 'no-undef'

Timer replied:
`no-undef` is a runtime error 99% of the time, the other 1% of the time doesn't make sense in a bundled system anyway (referencing `$` instead of `window.$` when using a script tag). To our build tool's knowledge, it will be a build error until the user alters the environment at runtime.

Timer replied:
`no-restricted-globals` is a tricky one, because it serves a few purposes -- first, many people are unaware of the entire list of globals, causing very confusing errors, e.g.:
```js
handleClick() {
  // other preface code
  event.target.value; // why is this blowing up ?!
}
```

Secondly, we want to keep a future of SSR applications open -- training people to use the `window.` prefix now will make this transition easier when running in a Node environment.

ttmarek replied:
> no-undef is a runtime error 99% of the time
Yeah sorry, I should have included this with the other run-time error eslint rules we have.

Timer replied:
`import/first` is arguably a style choice -- but it bites our users so often that I think an exclusion can be made here.

People try stuff like this all the time, and it looks like it would work:
```js
import $ from 'jquery';
window.jQuery = $;
import 'jquery-slimscroll'; // this module relies on jQuery being global
```

When in reality, that code is actually evaluated as:
```js
import $ from 'jquery';
import 'jquery-slimscroll';
window.jQuery = $;
```

In this case, you can argue it's not a style choice because they're not semantically identical.

Timer replied:
> If I understand correctly these are in place to help educate or train CRA users to follow certain best practices. But couldn't the same be said for every other eslint rule we have? I guess what I'm wondering, is how can we justify that these are set as errors while other eslint rules are set as warnings. Is it the severity of not following these practices? Or the difficultly of tracking bugs that could arise from not following these practices?

Hopefully my three explanations helped you understand our position a bit more. 😄 
I don't think other rules have such a strong argument.

ttmarek replied:
> Hopefully my three explanations helped you understand our position a bit more. :smile:
I don't think other rules have such a strong argument.

Kinda? Haha. Your explanations were helpful in understanding why `import/first` and `no-restricted-globals` are necessary, and why its a good idea to have them set as errors.

I guess what I'm still iffy on is the _criteria_ we're using to decide whether an eslint rule should be configured as a warning or error.

There are two criteria for making a rule an error that are crystal clear to me:
 
1. Run-time errors
2. Features that aren't supported
3. ???

What you said here is more what I was looking for:

> but it bites our users so often that I think an exclusion can be made here.

1. Run-time errors
2. Features that aren't supported
3. _stuff that causes our users so much grief that it's better that these are errors_?

Know what I mean? Sorry for being a pain in the ass. I hope y'all don't view this discussion as pedantic and unnecessary. It may help to have these criteria laid out though. In case others ask in the future.

Timer replied:
You're not a pain in the ass at all. 😄

At the end of the day, we do want to enforce what's best for majority of users.
So I feel like "stuff that causes our users so much grief that it's better that these are errors" is a pretty accurate description.

Perhaps more accurately, error for anything which has a non-obvious/indeterministic behavior to the majority of our users.

The `import` case requires a deep understanding of the spec and language features -- a lot of the community doesn't have this understanding yet (I even trip up on it occasionally).
Babel blurs this functionality, so it's probably going to remain one of those things we need to enforce even though advanced users know the caveats.

Maybe in a few years when `import`/`export` are mainstream in Node & browsers, and no longer require Babel we can remove the `import/first` rule.
But right now, it's so confusing because existing implementations don't follow spec / etc.

ttmarek commented:
@gaearon @Timer hope you guys are having a good fall. 🍂 🌧 

I've update the PR based on our discussion. As a quick refresher:

 - I removed eslint rules that are specific to formatting (basically any that collided with `eslint-config-prettier`.
 - I updated one of the comments in `packages/eslint-config-react-app/index.js` to clarify the criteria used to determine which eslint rules are configured as warnings and which rules are configured as errors.
 - I added more documentation to the User Guide around linting. It was a struggle for me understand CRA's approach to linting, I wanted to make it easier for the next guy.

Timer commented:
I'm on the fence about removing `no-mixed-operator`, just because we don't integrate with Prettier yet to make problems in logic statements more apparent.

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -163,7 +167,7 @@ Runs the app in the development mode.<br>
 
 The page will reload if you make edits.<br>
-You will also see any lint errors in the console.
+You will also see any lint warnings in the console.
```
Considering there may be lint errors or warnings, what about something like lint problems?
You can choose your word.

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -1408,12 +1375,157 @@ Example package.json:
+
+## Running the Linter
+
+Create React App includes a set of ESLint rules handpicked to help spot potential issues with your code and to enforce best practices. Create React App does not include any ESLint rules that enforce style or formatting. We recommend using a tool like [Prettier](https://github.com/prettier/prettier) for that instead. See the [Formatting Code Automatically](#formatting-code-automatically) section f ...
```
Maybe enforce best practices -> encourage best practices?

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -1408,12 +1375,157 @@ Example package.json:
+
+Create React App includes a set of ESLint rules handpicked to help spot potential issues with your code and to enforce best practices. Create React App does not include any ESLint rules that enforce style or formatting. We recommend using a tool like [Prettier](https://github.com/prettier/prettier) for that instead. See the [Formatting Code Automatically](#formatting-code-automatically) section f ...
+
+Create React App runs the linter automatically with the `npm start` and `npm build` commands. It's important to note that these two commands will only lint the files that are transpiled as part of the build, and any other code (e.g. tests) won't be linted.
```
I think we're starting to move away from the term "transpiled", we might as well take this as an opportunity to change this to "compiled".

ttmarek replied:
> move away from the term "transpiled"
Curious, why is that?

Timer replied:
I forget where it started, but it's been an initiative for a while.

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -1408,12 +1375,157 @@ Example package.json:
+
+### Customizing the Linter
+
+If you want to augment the ESLint rules bundled with Create React App, or if you want to run the linter on files that aren't transpiled as part of the build (e.g tests) you would need to create an `.eslintrc` file that extends `react-app` and place it in your project's root directory:
```
Ditto, transpiled -> compiled.

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -1408,12 +1375,157 @@ Example package.json:
-When creating a build of your application with `npm run build` linter warnings are not checked by default. Like `npm test`, you can force the build to perform a linter warning check by setting the environment variable `CI`. If any warnings are encountered then the build fails.
+This makes it obvious you are using a global variable intentionally rather than because of a typo.
+
+Alternatively, you can force the linter to ignore any line by adding `// eslint-disable-line` after it.
```
Hmm, I wonder why we called out this escape hatch here. Seems more general purpose that doesn't belong under the Global Variables section.

ttmarek replied:
I actually think this is okay. I guess we have a few options:

 - We can create a whole section on how to ignore the lint rules (basically [this](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments)).
   - Not a huge fan of this option. Mostly because we could just link to the eslint docs on this. It would also encourage CRA's users to disable eslint rules for entire blocks and files which may not be what's best for them, or even you, the CRA maintainers, as users will run into issues that you've worked to prevent by introducing the rules.
 - We can remove this line entirely.
    - This is a good option. It would encourage CRA's users not to disable eslint rules. One can argue that there may be cases where a rule doesn't make sense. But unlike other eslint rule sets (e.g. airbnb) I think there are very few (or no) rules y'all have selected that would ever make sense to disable. So if we follow our principles for selecting eslint rules, should CRA's users ever need to be educated about disabling eslint rules? Perhaps not.
 - We can leave this line in.
   - Not a bad option to have it hidden away here in the Global Variables section as a way to reference a global without the `window` prefix.

Timer replied:
Do we tell people how to ignore lint options anywhere else?

ttmarek replied:
Doesn't look like it no.

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -1408,12 +1375,157 @@ Example package.json:
+
+## Continuous Integration
+
+By default `npm test` runs the watcher with interactive CLI. However, you can force it to run tests once and finish the process by setting an environment variable called `CI`.
```
Let's be consistent with above and say setting the `CI` environment variable to `true`.
